### DISCUSSION TITLE
ProxyManager issues with JedisConnectionFactory from spring-data-redis

### DISCUSSION THREAD

**Discussion by airasyaqub** [vladimir-bukhtoyarov/bucket4j#426] [Q&A] [answered]
Hey @vladimir-bukhtoyarov 

**Redisson with Bucket4j** 

There are tons of tutorials our there to use bucket4j with redis. Almost all of them use redisson as client. They generally create a config & then using it create a CacheManager with a cache. Then they create a ProxyManager bean in which they pass that created cache using CacheManager bean. Following is the ProxyManager code which I guess is the relevant one for bucket4j 

```
  @Bean
  ProxyManager<String> proxyManager(CacheManager cacheManager) {
    return new JCacheProxyManager<>(cacheManager.getCache("cache_name"));
  }
```
  
 Here is one of the reference tutorial: https://www.freecodecamp.org/news/rate-limiting-with-bucket4j-and-redis/
 
 **Jedis with Bucket4j**
 While implementing the same flow with Jedis client I faced issues. I am using spring-data-redis package in my project which internally uses Jedis client. All of my redis needs such as DB, pub/sub are being fulfilled by it. In my proxyManager code I could not use the RedisCacheManager/CacheManager ( From spring-data-redis package ) in JCacheProxyManager because I have found out it requires JSR 107 specification internally & RedisCacheManager/CacheManager is not the compliant one. Redisson worked because it has JSR 107 [implementation](https://dzone.com/articles/jcache-api-jsr-107-implementation-for-redis) for it. 
 
 Then after much digging I have found some way to use Jedis with buckey4j. I fetch the already created JedisConnectionFactory bean ( which I am using for redis db, pub/sub ), get the jedis pool from it & create the JedisBasedProxyManager. Here is the code 
 
```
    @Bean
    ProxyManager<String> proxyManager(JedisConnectionFactory jedisConnectionFactory) throws NoSuchFieldException, IllegalAccessException {

        Field poolField = JedisConnectionFactory.class.getDeclaredField("pool");
        poolField.setAccessible(true);
        JedisPool pool = (JedisPool) poolField.get(jedisConnectionFactory);

        return JedisBasedProxyManager.builderFor(pool)
                .withExpirationStrategy(ExpirationAfterWriteStrategy.basedOnTimeForRefillingBucketUpToMax(ofSeconds(0)))
                .withKeyMapper(Mapper.STRING)
                .build();

    }
```
 **Issue with JedisBasedProxyManager**
 
 The issue is its not creating a hash in my redis cluster, what I want to do is some how use the cache created in RedisCacheManager bean in JedisBasedProxyManager or in any way create the hash in JedisBasedProxyManager. Here is the code where I am creating RedisCacheManager. 
 
```
   @Bean
    public CacheManager cacheManager(JedisConnectionFactory jedisConnectionFactory) {

        return RedisCacheManager.builder(jedisConnectionFactory)
                .cacheDefaults(RedisCacheConfiguration.defaultCacheConfig())
                .withInitialCacheConfigurations(Collections.singletonMap("MY_CACHE_NAME",
                        RedisCacheConfiguration.defaultCacheConfig().disableCachingNullValues()))
                .build();
    }
```

**Comment by vladimir-bukhtoyarov** [comment]
I do not recommend to work with Redis over JCache. This way is the infinite source of tricky problems.
Bucket4j provides directly integrations for tree major Java Redis libraries:
* Lettuci, see [this test code](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-redis/src/test/java/io/github/bucket4j/redis/RedisStandaloneTest.java#L100) as example of ProxyManager instatiation.
* Redisson, see [this test code](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-redis/src/test/java/io/github/bucket4j/redis/RedisStandaloneTest.java#L76) as example of ProxyManager instatiation.
* Jedis, see [this test code](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-redis/src/test/java/io/github/bucket4j/redis/RedisStandaloneTest.java#L123) as example of ProxyManager instatiation.


> The issue is its not creating a hash in my redis cluster

What does it mean?

> Field poolField = JedisConnectionFactory.class.getDeclaredField("pool");
poolField.setAccessible(true);
JedisPool pool = (JedisPool) poolField.get(jedisConnectionFactory);

JedisPool pool can be created without hacks:
```java
// for standalone Redis
new JedisPooled(redisHost, redisPort)
```
```java
// for Redis Cluster
List<HostAndPort> shards = new ArrayList<>();
shards.add(new HostAndPort(redisHost1, redisPort));
shards.add(new HostAndPort(redisHost2, redisPort));
shards.add(new HostAndPort(redisHost3, redisPort));
...
shards.add(new HostAndPort(redisHosttN, redisPort));
return new JedisCluster(new HashSet<>(shards));
```

**Comment by airasyaqub** [comment]
> I do not recommend to work with Redis over JCache

Then what distributed cache mechanism you suggest ? I believe Jcache does not have distributed cluster functionality. 

> The issue is its not creating a hash in my redis cluster

what I meant here is I would like to have capability of structuring my keys. For example if I am having an IP based rate limiting, so bucket key will be IP right, i.e key is 192.168.1.1, value will be bucket . How can I have key similar to my_rate_limit_cache:192.168.1.1, my_rate_limit_cache:192.168.2.2. Also what If I want to create a hash & include all my buckets in it ? How can I have control over these things. See

What key I am getting: 

<img width="1299" alt="Screenshot 2023-11-14 at 4 02 23 PM" src="https://github.com/bucket4j/bucket4j/assets/30695220/662a1aad-e1d3-421f-992b-75db27cee3db">

What key I need: 
<img width="1325" alt="Screenshot 2023-11-14 at 4 03 14 PM" src="https://github.com/bucket4j/bucket4j/assets/30695220/c5a5116d-8b08-4b66-a425-0f8df2ac237e">


> JedisPool pool can be created without hacks:

The reason I did this reflection API hack is because I wanted to use the same jedispool created in my JedisConnectionFactory. I did not want to create new one. Its not a big deal but still whats your opinion about it ?

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
> what I meant here is I would like to have capability of structuring my keys.

But... Redis hash does not provide expiration for individual items in the hash, it makes impossible to expire buckets that became unused.  Redisson integration had been initially written on top of Redisson/RMap that uses RedisHash behind, but then it was rewritten to use `org.redisson.command.CommandExecutor` in order to add TTL feature, you can see discussion there https://github.com/bucket4j/bucket4j/discussions/161#discussioncomment-1165086

> Then what distributed cache mechanism you suggest ?

Feel free to stay with Redis, but use one of direct Bucket4j integration with Jedis, Lettuce or Redisson. But if organizing multiple buckets in one collection is dominant requirement then I would recommend to switch to ApacheIgnite/Hazelcast/Infinispan.

**Reply by airasyaqub** [reply]
Yeah I think the current Jedis one I am using is fine. It just a grouping thing & I can sacrifice that. Thanks @vladimir-bukhtoyarov
